src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs

```diff
@@ -146,7 +146,18 @@ public RestoreCommand(RestoreRequest request)
             _success = !request.AdditionalMessages?.Any(m => m.Level == LogLevel.Error) ?? true;
 
             // Enable the new dependency resolver if the project is using PackageReference, transitive pinning is disabled, and the user has not explicitly opted out of using it
-            _enableNewDependencyResolver = request.Project.RestoreMetadata.ProjectStyle == ProjectStyle.PackageReference && !_request.Project.RestoreMetadata.CentralPackageTransitivePinningEnabled && !_request.Project.RestoreMetadata.UseLegacyDependencyResolver;
+
+            if (request.Project.RestoreMetadata.ProjectStyle == ProjectStyle.PackageReference
+                && !_request.Project.RestoreMetadata.CentralPackageTransitivePinningEnabled
+                && !_request.Project.RestoreMetadata.UseLegacyDependencyResolver)
+            {
+                _enableNewDependencyResolver = true;
+            }
+            else
+            {
+                _enableNewDependencyResolver = false;
+                _request.Project.RestoreMetadata.UseLegacyDependencyResolver = true;
```

Comment:
The PackageSpec should not be edited in the restore command. It should really be treated as immutable/
If you look at the telemetry I added, it differentiates between property specified and used and legacy algorithm used because the new resolver doesn't support a certain scenario.

Reply:
Without this, I have to update a few dozen tests because they only set the `CentralPackageTransitivePinningEnabled` property and not the `UseLegacyDependencyResolver` property.  This seemed cleaner to me than having to update all of those tests.

When I get transitive pinning working for the new resolver, I should be able to remove this line.

Reply:
Changing the package spec can mess with no-op of related projects.
We probably don't have tests like that, but it's how we've been able to reduce the amount of package spec cloning we do.

Reply:
Basically https://github.com/NuGet/NuGet.Client/pull/5930 is what we'll want eventually.

Reply:
> Without this, I have to update a few dozen tests because they only set the CentralPackageTransitivePinningEnabled property and not the UseLegacyDependencyResolver property. This seemed cleaner to me than having to update all of those tests.


Can PackageSpecReferenceDependencyProvider repeat the same check as the one here?

Reply:
Technically yes, it just places the logic in two places which was how it was before so I had changed it to read this property instead.

Reply:
So I expect this to be temporary until we fix the transitive pinning correct?

If that's the case I'd rather duplicate some code, than risk breaking no-op and cause more restores.

Reply:
> Changing the package spec can mess with no-op of related projects.

In general, yes. In this case, it's being mutated in `RestoreCommand`'s constructor, which happens before no-op check (ExecuteAsync?). Personally, I don't care if the first restore after a VS/.NET upgrade fails no-op, only that for the same version of NuGet with the same project inputs has working no-op.

I also completely agree that PackageSpec and DGSpec should be immutable, but even if that was already implemented, this code could have done something like:

```cs
var newRestoreMetadata = new RestoreMetadata(request.Project.RestoreMetadata) { UseLegacy = true} ;
var newProject = new Project(request.Project) { RestoreMetadata = newRestoreMetadata };
_request = new RestoreRequest(request) { Project = newProject };
```

Again, this is in the constructor, before the restore actually starts. Generally agree that our code shouldn't do this, we should design Nuget to deserialize the project inputs into a PackageSpec and then never modify or clone/copy it.

I wish that PackageSpecReferenceDependencyProvider was in NuGet.Commands, not NuGet.ProjectModel (which in my mind should only has DTO objects needed for the DGSpec and maybe APIs to (de)serialize it). In which case we could have an internal helper function and call it in two places, but that's not code we have, so either we duplicate business logic (I have no hesitation in duplicate boilerplate code, but business logic is where Don't Repeat Yourself reduces bugs)

But at the same time, it's temporary until the new resolver implements the feature, and it's a simple if statement that needs to be duplicated.  I'm unconvinced that there's a best solution here. I'm just sharing my thoughts, in case anyone cares.  This PR looks good to me, and I'd be happy to approve once this conversation is resolved.

Reply:
PackageSpecReferenceDependencyProvider has the PackageSpec so it can check the exact same condition.

Reply:
I've removed the setting of this for now.